File: cpp/tests/unit/test_bugs_new5.cpp
// TDD failing tests for BUG-NEW-24, BUG-NEW-25, BUG-NEW-26, BUG-NEW-27 (C++ side).
//
// Each test is written to FAIL before the fix and PASS after the fix is applied.
// Positive/regression tests that should already PASS are explicitly noted.
//
// ─────────────────────────────────────────────────────────────────────────────
// BUG-NEW-24 (C++ only, §4.4.2.8/9/10):
//   CMD_TLBI_EL2_VA, CMD_TLBI_EL2_VAA, CMD_TLBI_EL2_ASID are missing the
//   IDR0.Hyp==0 → CERROR_ILL guard.  They fall through into the shared block
//   at smmu.cpp lines 4388-4390 with no guard.  Only CMD_TLBI_EL2_ALL (separate
//   case) already has the guard.
//   Per ARM §4.4.2.8/9/10: "When SMMU_IDR0.Hyp == 0, this command causes
//   CERROR_ILL."
//
// BUG-NEW-25 (C++ side, §4.2.1/4.2.2/4.1.6):
//   CMD_PREFETCH_CONFIG and CMD_PREFETCH_ADDR do not check SSec=1 and therefore
//   do not raise CERROR_ILL.  All NS-queue commands with SSec=1 must raise
//   CERROR_ILL per §4.1.6.
//
// BUG-NEW-26 (C++ only, §4.7.3/4.8):
//   CMD_SYNC completion event inherits stream security state via a stream lookup.
//   CMD_SYNC has no architectural StreamID; the event must always be NonSecure.
//   In C++ (smmu.cpp lines ~4545-4565), cs != 0 triggers a stream lookup which
//   can yield Secure.  The fix: always use SecurityState::NonSecure.
// ─────────────────────────────────────────────────────────────────────────────

#include <gtest/gtest.h>
#include "smmu/smmu.h"
#include "smmu/types.h"
#include "smmu/configuration.h"
#include <cstdint>
#include <vector>

using namespace smmu;

// ============================================================================
// Helpers
// ============================================================================

namespace {

static void enableSMMU(SMMU& s) {
    s.setCR0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN | SMMU::CR0_PRIQEN);
}

// Build a minimal CommandEntry for the given type.
static CommandEntry makeCmd(CommandType type) {
    CommandEntry cmd;
    cmd.type  = type;
    cmd.cs    = 0u;
    cmd.ssec  = false;
    return cmd;
}

// Check whether GERROR.CMDQ_ERR is active (GERROR XOR GERRORN has bit 0 set).
// BUG-NEW-40 fix: getGerror() already returns gerrorStatus XOR gerrorNStatus
// (the active error bits).  The former double-XOR recovered the raw gerrorStatus
// instead of the active-error mask.  One mask is sufficient.
static bool isGerrorCmdqErrActive(const SMMU& s) {
    return (s.getGerror() & GERROR_CMDQ_ERR) != 0u;
}

} // anonymous namespace

// ============================================================================
// BUG-NEW-24: CMD_TLBI_EL2_VA/VAA/ASID must check IDR0.Hyp (§4.4.2.8/9/10)
// ============================================================================
// ARM §4.4.2.8/9/10: "When SMMU_IDR0.Hyp == 0, this command causes CERROR_ILL."
// The fix breaks TLBI_EL2_VA, TLBI_EL2_VAA, and TLBI_EL2_ASID out of the
// shared fall-through block and adds an IDR0.Hyp guard for each.

// Test 1: When Hyp=false (setHypSupported(false)), CMD_TLBI_EL2_VA must
// activate GERROR.CMDQ_ERR and write CERROR_ILL.
//
// BEFORE FIX: no guard → command executes without error → CMDQ_ERR stays 0
//             → EXPECT_TRUE(isGerrorCmdqErrActive) FAILS (red).
// AFTER FIX:  guard fires → CERROR_ILL written → CMDQ_ERR toggled → passes.
TEST(BugNew24El2VaVaaAsidHypGuard, El2VA_Missing_IDR0Hyp_Guard) {
    SMMU smmu;
    enableSMMU(smmu);

    // Override: clear IDR0.Hyp to simulate a non-hypervisor implementation.
    smmu.setHypSupported(false);
    ASSERT_EQ(smmu.getIDR0() & (1u << 9u), 0u)
        << "BUG-NEW-24 precondition: IDR0.Hyp (bit 9) must be 0 after setHypSupported(false)";

    CommandEntry cmd = makeCmd(CommandType::TLBI_EL2_VA);
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_ILL)
        << "BUG-NEW-24: CMD_TLBI_EL2_VA with IDR0.Hyp=0 must write CERROR_ILL "
           "(ARM §4.4.2.8)";
    EXPECT_TRUE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-24: CMD_TLBI_EL2_VA with IDR0.Hyp=0 must assert GERROR.CMDQ_ERR "
           "(ARM §4.4.2.8)";
}

// Test 2: When Hyp=false, CMD_TLBI_EL2_VAA must activate GERROR.CMDQ_ERR.
//
// BEFORE FIX: no guard → no CERROR → test FAILS (red).
// AFTER FIX:  guard fires → CERROR_ILL and CMDQ_ERR set → passes.
TEST(BugNew24El2VaVaaAsidHypGuard, El2VAA_Missing_IDR0Hyp_Guard) {
    SMMU smmu;
    enableSMMU(smmu);

    smmu.setHypSupported(false);
    ASSERT_EQ(smmu.getIDR0() & (1u << 9u), 0u)
        << "BUG-NEW-24 precondition: IDR0.Hyp must be 0";

    CommandEntry cmd = makeCmd(CommandType::TLBI_EL2_VAA);
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_ILL)
        << "BUG-NEW-24: CMD_TLBI_EL2_VAA with IDR0.Hyp=0 must write CERROR_ILL "
           "(ARM §4.4.2.9)";
    EXPECT_TRUE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-24: CMD_TLBI_EL2_VAA with IDR0.Hyp=0 must assert GERROR.CMDQ_ERR "
           "(ARM §4.4.2.9)";
}

// Test 3: When Hyp=false, CMD_TLBI_EL2_ASID must activate GERROR.CMDQ_ERR.
//
// BEFORE FIX: no guard → no CERROR → test FAILS (red).
// AFTER FIX:  guard fires → CERROR_ILL and CMDQ_ERR set → passes.
TEST(BugNew24El2VaVaaAsidHypGuard, El2ASID_Missing_IDR0Hyp_Guard) {
    SMMU smmu;
    enableSMMU(smmu);

    smmu.setHypSupported(false);
    ASSERT_EQ(smmu.getIDR0() & (1u << 9u), 0u)
        << "BUG-NEW-24 precondition: IDR0.Hyp must be 0";

    CommandEntry cmd = makeCmd(CommandType::TLBI_EL2_ASID);
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_ILL)
        << "BUG-NEW-24: CMD_TLBI_EL2_ASID with IDR0.Hyp=0 must write CERROR_ILL "
           "(ARM §4.4.2.10)";
    EXPECT_TRUE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-24: CMD_TLBI_EL2_ASID with IDR0.Hyp=0 must assert GERROR.CMDQ_ERR "
           "(ARM §4.4.2.10)";
}

// Test 4 (positive): When Hyp=true (default), CMD_TLBI_EL2_VA must NOT raise
// CERROR_ILL.
//
// This positive path must PASS both before and after the fix (regression guard).
TEST(BugNew24El2VaVaaAsidHypGuard, Hyp1_EL2VA_Succeeds) {
    SMMU smmu;
    enableSMMU(smmu);

    // IDR0.Hyp defaults to true (bit 9 = 1).
    ASSERT_NE(smmu.getIDR0() & (1u << 9u), 0u)
        << "BUG-NEW-24 precondition: IDR0.Hyp (bit 9) must be 1 by default";

    CommandEntry cmd = makeCmd(CommandType::TLBI_EL2_VA);
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_NONE)
        << "BUG-NEW-24 (positive): CMD_TLBI_EL2_VA with IDR0.Hyp=1 must NOT "
           "write CERROR_ILL (ARM §4.4.2.8)";
    EXPECT_FALSE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-24 (positive): CMD_TLBI_EL2_VA with IDR0.Hyp=1 must NOT "
           "assert GERROR.CMDQ_ERR (ARM §4.4.2.8)";
}

// ============================================================================
// BUG-NEW-25 (C++): CMD_PREFETCH_CONFIG and CMD_PREFETCH_ADDR with SSec=1
// must raise CERROR_ILL (§4.1.6: all NS-queue commands with SSec=1 are illegal)
// ============================================================================

// Test 5: CMD_PREFETCH_CONFIG with ssec=true must raise CERROR_ILL.
//
// BEFORE FIX: both commands are unconditional no-ops at smmu.cpp lines 4337-4345
//             with no ssec check → CMDQ_ERR not set → test FAILS (red).
// AFTER FIX:  ssec guard added → CERROR_ILL written → CMDQ_ERR toggled → passes.
TEST(BugNew25PrefetchSsecGuard, PrefetchConfig_SSec1_Must_Raise_CERROR_ILL) {
    SMMU smmu;
    enableSMMU(smmu);

    CommandEntry cmd = makeCmd(CommandType::PREFETCH_CONFIG);
    cmd.ssec = true;
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_ILL)
        << "BUG-NEW-25: CMD_PREFETCH_CONFIG with ssec=1 must write CERROR_ILL "
           "(ARM §4.1.6)";
    EXPECT_TRUE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-25: CMD_PREFETCH_CONFIG with ssec=1 must assert GERROR.CMDQ_ERR "
           "(ARM §4.1.6)";
}

// Test 6: CMD_PREFETCH_ADDR with ssec=true must raise CERROR_ILL.
//
// BEFORE FIX: unconditional no-op → no CERROR → test FAILS (red).
// AFTER FIX:  ssec guard fires → CERROR_ILL and CMDQ_ERR set → passes.
TEST(BugNew25PrefetchSsecGuard, PrefetchAddr_SSec1_Must_Raise_CERROR_ILL) {
    SMMU smmu;
    enableSMMU(smmu);

    CommandEntry cmd = makeCmd(CommandType::PREFETCH_ADDR);
    cmd.ssec = true;
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_ILL)
        << "BUG-NEW-25: CMD_PREFETCH_ADDR with ssec=1 must write CERROR_ILL "
           "(ARM §4.1.6)";
    EXPECT_TRUE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-25: CMD_PREFETCH_ADDR with ssec=1 must assert GERROR.CMDQ_ERR "
           "(ARM §4.1.6)";
}

// Test 7 (positive): CMD_PREFETCH_CONFIG with ssec=false must NOT raise error.
//
// This positive path must PASS both before and after the fix (regression guard).
TEST(BugNew25PrefetchSsecGuard, PrefetchConfig_SSec0_NoError) {
    SMMU smmu;
    enableSMMU(smmu);

    CommandEntry cmd = makeCmd(CommandType::PREFETCH_CONFIG);
    cmd.ssec = false;
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    EXPECT_EQ(smmu.getCmdqConsErr(), CERROR_NONE)
        << "BUG-NEW-25 (positive): CMD_PREFETCH_CONFIG with ssec=0 must NOT "
           "write CERROR_ILL (ARM §4.1.6)";
    EXPECT_FALSE(isGerrorCmdqErrActive(smmu))
        << "BUG-NEW-25 (positive): CMD_PREFETCH_CONFIG with ssec=0 must NOT "
           "assert GERROR.CMDQ_ERR (ARM §4.1.6)";
}

// ============================================================================
// BUG-NEW-26 (C++): CMD_SYNC completion event must always use NonSecure
// ============================================================================
// ARM §4.7.3/§4.8: CMD_SYNC has no StreamID operand; the completion event's
// security state must always be SecurityState::NonSecure.
// In C++ (smmu.cpp lines ~4545-4565), when cs != 0 the code looks up the
// stream by command.streamID and uses that stream's security state.
// If a Secure stream has been configured for that StreamID, the event
// incorrectly inherits Secure.

// Test 8: Configure a Secure stream; issue CMD_SYNC with cs=1 pointing to
// that stream's StreamID.  The COMMAND_SYNC_COMPLETION event must have
// securityState == NonSecure.
//
// BEFORE FIX: stream lookup returns Secure → event.securityState=Secure
//             → EXPECT_EQ(Secure, NonSecure) FAILS (red).
// AFTER FIX:  stream lookup removed, always NonSecure → event OK → passes.
TEST(BugNew26CmdSyncSecState, CompletionEvent_AlwaysNonSecure) {
    SMMU smmu;
    enableSMMU(smmu);

    // Configure and enable a Secure stream so the stream-lookup code can find
    // it with isStreamEnabled() == true.  Without enableStream() the C++ code
    // falls back to NonSecure unconditionally, masking the bug.
    StreamConfig cfg;
    cfg.translationEnabled = true;
    cfg.stage1Enabled      = true;
    cfg.stage2Enabled      = false;
    cfg.securityState      = SecurityState::Secure;
    cfg.strw               = StreamWorld::EL1_EL0;
    smmu.configureStream(0x60u, cfg);
    smmu.enableStream(0x60u);

    // CMD_SYNC with cs=1 (SIG_IRQ) to trigger a COMMAND_SYNC_COMPLETION event.
    // stream_id=0x60 points at the Secure stream; the event must ignore this.
    CommandEntry cmd = makeCmd(CommandType::SYNC);
    cmd.cs       = 1u;
    cmd.streamID = 0x60u;
    smmu.submitCommand(cmd);
    smmu.processCommandQueue();

    const std::vector<EventEntry> events = smmu.getEventQueue();
    bool found = false;
    for (const auto& ev : events) {
        if (ev.type == EventType::COMMAND_SYNC_COMPLETION) {
            EXPECT_EQ(ev.securityState, SecurityState::NonSecure)
                << "BUG-NEW-26: COMMAND_SYNC_COMPLETION must always use "
                   "SecurityState::NonSecure — CMD_SYNC has no StreamID "
                   "operand; the completion event belongs to the NS command "
                   "queue domain (ARM §4.7.3/§4.8)";
            found = true;
        }
    }
    EXPECT_TRUE(found)
        << "BUG-NEW-26: COMMAND_SYNC_COMPLETION event must be generated for CS=1";
}

File: rust/smmu/tests/test_bugs_new5.rs
//! TDD failing tests for BUG-NEW-27 and BUG-NEW-25 (Rust side).
//!
//! Each test is written to FAIL with the current code and PASS only after the
//! corresponding fix is applied.  Positive/regression tests that should
//! already PASS are explicitly noted.
//!
//! # Bug Summary
//!
//! ## BUG-NEW-27 (Rust, §4.4.2.8/9/10)
//!
//! `TlbiEl2Va` is grouped with `TlbiNhVa` and `TlbiSEl2Va` in a single match
//! arm (line ~5641 of smmu/mod.rs) with no IDR0.Hyp guard.
//! `TlbiEl2Vaa` is grouped with `TlbiNhVaa` and `TlbiSEl2Vaa` (line ~5683)
//! with no IDR0.Hyp guard.
//! `TlbiEl2Asid` (line ~5593) also has no IDR0.Hyp guard.
//!
//! ARM §4.4.2.8/9/10: "When SMMU_IDR0.Hyp == 0, this command causes CERROR_ILL."
//!
//! BEFORE FIX: Hyp=false does not prevent execution → CMDQ_ERR not set → test FAILS.
//! AFTER FIX:  IDR0.Hyp==0 guard added to each arm → CERROR_ILL raised → passes.
//!
//! ## BUG-NEW-25 (Rust, §4.1.6/§4.2.1/§4.2.2)
//!
//! `CommandType::PrefetchConfig | CommandType::PrefetchAddr` at line ~6035 of
//! smmu/mod.rs is an unconditional no-op with no `ssec` check.
//!
//! ARM §4.1.6: ALL NS-queue commands with SSec=1 must raise CERROR_ILL.
//!
//! BEFORE FIX: no ssec check → no CERROR → test FAILS.
//! AFTER FIX:  ssec guard added → CERROR_ILL raised → passes.
#![allow(missing_docs)]
#![allow(clippy::unwrap_used)]
#![allow(clippy::items_after_statements)]
#![allow(clippy::doc_markdown)]
#![allow(clippy::uninlined_format_args)]

use smmu::types::{CommandEntry, CommandType, SecurityState};
use smmu::SMMU;

// ============================================================================
// Helpers
// ============================================================================

/// Build an SMMU with SMMUEN + CMDQEN + EVENTQEN + PRIQEN enabled.
fn make_smmu() -> SMMU {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_CMDQEN | SMMU::CR0_EVENTQEN | SMMU::CR0_PRIQEN);
    smmu
}

/// Returns true when GERROR.CMDQ_ERR is active (GERROR XOR GERRORN has bit 0 set).
fn is_gerror_cmdq_err_active(smmu: &SMMU) -> bool {
    (smmu.get_gerror() ^ smmu.get_gerrorn()) & SMMU::GERROR_CMDQ_ERR != 0
}

// ============================================================================
// BUG-NEW-27: TlbiEl2Va/TlbiEl2Vaa/TlbiEl2Asid must check IDR0.Hyp
// ============================================================================
// ARM §4.4.2.8/9/10: "When SMMU_IDR0.Hyp == 0, this command causes CERROR_ILL."
// The three EL2-targeted TLBI commands need their own match arms with an
// IDR0.Hyp guard, separated from the TlbiNhVa/TlbiNhVaa/TlbiNhAsid peers
// that have no such restriction.

/// BUG-NEW-27 (primary): with `set_hyp_supported(false)`, submitting
/// `TlbiEl2Va` must set CMDQ_ERR and write CERROR_ILL.
///
/// ARM §4.4.2.8: "If SMMU_IDR0.Hyp==0, CERROR_ILL."
///
/// BEFORE FIX: `TlbiEl2Va` is grouped with `TlbiNhVa` — no guard fires →
///   `is_gerror_cmdq_err_active` returns false → test FAILS (red).
/// AFTER FIX:  dedicated arm with Hyp guard raises CERROR_ILL → passes (green).
#[test]
fn bug_new27_tlbi_el2_va_missing_hyp_guard() {
    let smmu = make_smmu();

    // Override IDR0.Hyp to simulate a non-hypervisor implementation.
    smmu.set_hyp_supported(false);
    assert_eq!(
        smmu.get_idr0() & (1 << 9),
        0,
        "BUG-NEW-27 precondition: IDR0.Hyp (bit 9) must be 0 after set_hyp_supported(false)"
    );

    let cmd = CommandEntry::new(CommandType::TlbiEl2Va, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-27: TlbiEl2Va with IDR0.Hyp=0 must write CERROR_ILL \
         (ARM §4.4.2.8). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-27: TlbiEl2Va with IDR0.Hyp=0 must assert GERROR.CMDQ_ERR \
         (ARM §4.4.2.8)"
    );
}

/// BUG-NEW-27: with `set_hyp_supported(false)`, `TlbiEl2Vaa` must
/// set CMDQ_ERR and write CERROR_ILL.
///
/// ARM §4.4.2.9: "If SMMU_IDR0.Hyp==0, CERROR_ILL."
///
/// BEFORE FIX: `TlbiEl2Vaa` grouped with `TlbiNhVaa` — no guard → test FAILS.
/// AFTER FIX:  dedicated arm with Hyp guard raises CERROR_ILL → passes.
#[test]
fn bug_new27_tlbi_el2_vaa_missing_hyp_guard() {
    let smmu = make_smmu();

    smmu.set_hyp_supported(false);
    assert_eq!(
        smmu.get_idr0() & (1 << 9),
        0,
        "BUG-NEW-27 precondition: IDR0.Hyp must be 0"
    );

    let cmd = CommandEntry::new(CommandType::TlbiEl2Vaa, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-27: TlbiEl2Vaa with IDR0.Hyp=0 must write CERROR_ILL \
         (ARM §4.4.2.9). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-27: TlbiEl2Vaa with IDR0.Hyp=0 must assert GERROR.CMDQ_ERR \
         (ARM §4.4.2.9)"
    );
}

/// BUG-NEW-27: with `set_hyp_supported(false)`, `TlbiEl2Asid` must
/// set CMDQ_ERR and write CERROR_ILL.
///
/// ARM §4.4.2.10: "If SMMU_IDR0.Hyp==0, CERROR_ILL."
///
/// BEFORE FIX: `TlbiEl2Asid` has its own arm but no Hyp guard → test FAILS.
/// AFTER FIX:  Hyp guard added → CERROR_ILL raised → passes.
#[test]
fn bug_new27_tlbi_el2_asid_missing_hyp_guard() {
    let smmu = make_smmu();

    smmu.set_hyp_supported(false);
    assert_eq!(
        smmu.get_idr0() & (1 << 9),
        0,
        "BUG-NEW-27 precondition: IDR0.Hyp must be 0"
    );

    let cmd = CommandEntry::new(CommandType::TlbiEl2Asid, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-27: TlbiEl2Asid with IDR0.Hyp=0 must write CERROR_ILL \
         (ARM §4.4.2.10). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-27: TlbiEl2Asid with IDR0.Hyp=0 must assert GERROR.CMDQ_ERR \
         (ARM §4.4.2.10)"
    );
}

/// BUG-NEW-27 (positive): with the default IDR0.Hyp=1, `TlbiEl2Va` must
/// execute normally — no CERROR_ILL, no GERROR.CMDQ_ERR.
///
/// This test verifies that the IDR0.Hyp guard does NOT fire when Hyp is
/// supported (the common/default case).
///
/// BEFORE FIX: passes trivially (no guard at all).
/// AFTER FIX:  guard present but Hyp=1 → no CERROR_ILL → still passes.
#[test]
fn bug_new27_hyp1_el2_va_succeeds() {
    let smmu = make_smmu();

    // IDR0.Hyp defaults to true (bit 9 = 1).
    let idr0 = smmu.get_idr0();
    assert!(
        (idr0 & (1 << 9)) != 0,
        "BUG-NEW-27 precondition: IDR0.Hyp (bit 9) must be 1 by default. \
         got IDR0=0x{:08x}",
        idr0
    );

    let cmd = CommandEntry::new(CommandType::TlbiEl2Va, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let result = smmu.process_command_queue();

    assert!(
        result.is_ok(),
        "BUG-NEW-27 (positive): TlbiEl2Va with IDR0.Hyp=1 must succeed. Got {:?}",
        result
    );
    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_NONE,
        "BUG-NEW-27 (positive): TlbiEl2Va with IDR0.Hyp=1 must not set CERROR_ILL \
         (ARM §4.4.2.8)"
    );
    assert!(
        !is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-27 (positive): TlbiEl2Va with IDR0.Hyp=1 must not assert GERROR.CMDQ_ERR \
         (ARM §4.4.2.8)"
    );
}

// ============================================================================
// BUG-NEW-25 (Rust): PrefetchConfig and PrefetchAddr with ssec=true
// must raise CERROR_ILL (§4.1.6: all NS-queue commands with SSec=1 illegal)
// ============================================================================

/// BUG-NEW-25 (primary): `PrefetchConfig` with `ssec=true` must set
/// CMDQ_ERR and write CERROR_ILL.
///
/// ARM §4.1.6: "Any NS-queue command with SSec=1 is illegal — CERROR_ILL."
///
/// BEFORE FIX: `PrefetchConfig | PrefetchAddr => {}` (unconditional no-op,
///   no ssec check) → CMDQ_ERR not set → test FAILS (red).
/// AFTER FIX:  ssec guard added before the no-op branch → CERROR_ILL raised
///   → passes (green).
#[test]
fn bug_new25_prefetch_config_ssec1_cerror_ill() {
    let smmu = make_smmu();

    let mut cmd = CommandEntry::new(CommandType::PrefetchConfig, 0, 0);
    cmd.ssec = true;
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-25: PrefetchConfig with ssec=1 must write CERROR_ILL \
         (ARM §4.1.6). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-25: PrefetchConfig with ssec=1 must assert GERROR.CMDQ_ERR \
         (ARM §4.1.6)"
    );
}

/// BUG-NEW-25: `PrefetchAddr` with `ssec=true` must set CMDQ_ERR and
/// write CERROR_ILL.
///
/// BEFORE FIX: unconditional no-op → no CERROR → test FAILS (red).
/// AFTER FIX:  ssec guard fires → CERROR_ILL raised → passes (green).
#[test]
fn bug_new25_prefetch_addr_ssec1_cerror_ill() {
    let smmu = make_smmu();

    let mut cmd = CommandEntry::new(CommandType::PrefetchAddr, 0, 0);
    cmd.ssec = true;
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-25: PrefetchAddr with ssec=1 must write CERROR_ILL \
         (ARM §4.1.6). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-25: PrefetchAddr with ssec=1 must assert GERROR.CMDQ_ERR \
         (ARM §4.1.6)"
    );
}

/// BUG-NEW-25 (positive): `PrefetchConfig` with `ssec=false` must NOT
/// raise any error.
///
/// This positive path must PASS both before and after the fix (regression guard).
#[test]
fn bug_new25_prefetch_config_ssec0_no_error() {
    let smmu = make_smmu();

    let mut cmd = CommandEntry::new(CommandType::PrefetchConfig, 0, 0);
    cmd.ssec = false;
    smmu.submit_command(cmd).unwrap();
    let result = smmu.process_command_queue();

    assert!(
        result.is_ok(),
        "BUG-NEW-25 (positive): PrefetchConfig with ssec=0 must succeed. Got {:?}",
        result
    );
    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_NONE,
        "BUG-NEW-25 (positive): PrefetchConfig with ssec=0 must not set CERROR_ILL \
         (ARM §4.1.6)"
    );
    assert!(
        !is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-25 (positive): PrefetchConfig with ssec=0 must not assert \
         GERROR.CMDQ_ERR (ARM §4.1.6)"
    );
}

// ============================================================================
// Additional regression: BUG-NEW-25 PrefetchAddr ssec=false (no error)
// ============================================================================

/// BUG-NEW-25 (positive): `PrefetchAddr` with `ssec=false` must NOT raise
/// any error.
///
/// Regression guard to ensure the ssec=false path still works after the fix.
#[test]
fn bug_new25_prefetch_addr_ssec0_no_error() {
    let smmu = make_smmu();

    let mut cmd = CommandEntry::new(CommandType::PrefetchAddr, 0, 0);
    cmd.ssec = false;
    smmu.submit_command(cmd).unwrap();
    let result = smmu.process_command_queue();

    assert!(
        result.is_ok(),
        "BUG-NEW-25 (positive): PrefetchAddr with ssec=0 must succeed. Got {:?}",
        result
    );
    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_NONE,
        "BUG-NEW-25 (positive): PrefetchAddr with ssec=0 must not set CERROR_ILL \
         (ARM §4.1.6)"
    );
    assert!(
        !is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-25 (positive): PrefetchAddr with ssec=0 must not assert \
         GERROR.CMDQ_ERR (ARM §4.1.6)"
    );
}

// ============================================================================
// Additional BUG-NEW-27 regression: TlbiEl2Vaa and TlbiEl2Asid with Hyp=true
// ============================================================================

/// BUG-NEW-27 (positive): `TlbiEl2Vaa` with IDR0.Hyp=1 must NOT raise CERROR_ILL.
#[test]
fn bug_new27_hyp1_el2_vaa_succeeds() {
    let smmu = make_smmu();

    assert!(
        (smmu.get_idr0() & (1 << 9)) != 0,
        "BUG-NEW-27 precondition: IDR0.Hyp must be 1 by default"
    );

    let cmd = CommandEntry::new(CommandType::TlbiEl2Vaa, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let result = smmu.process_command_queue();

    assert!(
        result.is_ok(),
        "BUG-NEW-27 (positive): TlbiEl2Vaa with IDR0.Hyp=1 must succeed. Got {:?}",
        result
    );
    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_NONE,
        "BUG-NEW-27 (positive): TlbiEl2Vaa with IDR0.Hyp=1 must not set CERROR_ILL"
    );
}

/// BUG-NEW-27 (positive): `TlbiEl2Asid` with IDR0.Hyp=1 must NOT raise CERROR_ILL.
#[test]
fn bug_new27_hyp1_el2_asid_succeeds() {
    let smmu = make_smmu();

    assert!(
        (smmu.get_idr0() & (1 << 9)) != 0,
        "BUG-NEW-27 precondition: IDR0.Hyp must be 1 by default"
    );

    let cmd = CommandEntry::new(CommandType::TlbiEl2Asid, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let result = smmu.process_command_queue();

    assert!(
        result.is_ok(),
        "BUG-NEW-27 (positive): TlbiEl2Asid with IDR0.Hyp=1 must succeed. Got {:?}",
        result
    );
    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_NONE,
        "BUG-NEW-27 (positive): TlbiEl2Asid with IDR0.Hyp=1 must not set CERROR_ILL"
    );
}

// ============================================================================
// Additional BUG-NEW-26 regression (Rust side)
// ============================================================================
// BUG-NEW-26 was already fixed in Rust (test_bugs_new4.rs covers it), but we
// add a cross-reference guard here to ensure the Rust side stays correct.

/// BUG-NEW-26 (regression): Rust CMD_SYNC completion event must be NonSecure
/// regardless of stream security state.
///
/// This test verifies the already-fixed Rust path remains correct as a
/// regression guard alongside the new BUG-NEW-27 / BUG-NEW-25 tests.
#[test]
fn bug_new26_cmd_sync_completion_nonsecure_regression() {
    use smmu::types::{StreamConfig, StreamID};

    let smmu = make_smmu();

    let stream_id = StreamID::new(0xC0).unwrap();
    let secure_config = StreamConfig {
        security_state: SecurityState::Secure,
        ..StreamConfig::stage1_only()
    };
    smmu.configure_stream(stream_id, secure_config).unwrap();

    let mut cmd = CommandEntry::new(CommandType::Sync, 0xC0, 0);
    cmd.cs = 0x01; // SIG_IRQ → emits CommandSyncCompletion

    smmu.submit_command(cmd).expect("submit_command must succeed");
    smmu.process_command_queue().expect("process_command_queue must succeed");

    let events = smmu.get_events();
    use smmu::types::EventType;
    let completion = events
        .iter()
        .find(|e| e.event_type == EventType::CommandSyncCompletion)
        .expect("CommandSyncCompletion event must be present");

    assert_eq!(
        completion.security_state,
        SecurityState::NonSecure,
        "BUG-NEW-26 (regression): CommandSyncCompletion.security_state must always \
         be NonSecure. Got: {:?}",
        completion.security_state
    );
}
